### PR TITLE
[AA] fix UserOpMethodHandler estimateUserOperationGas #752

### LINKED ISSUE #752: [AA] fix UserOpMethodHandler estimateUserOperationGas
In `UserOpMethodHandler.ts`
```
// todo: validation manager duplicate?
    const errorResult = await this.entryPoint.callStatic
      .simulateValidation(userOp)
      .catch((e) => e)
```
This needs to be routed through `EntryPointWrapper` because of Solidity custom errors.

This currently works (in unit tests) because `UserOpMethodHandler.test.ts` is run against Hardhat. Adapt the tests.

Integration tests don't actually use Bundler RPC method `eth_estimateUserOperationGas` - we should make note in the docs this currently isn't supported.

### PR DESCRIPTION
**Pull request by wsdt**
## Overview

[AA] fix UserOpMethodHandler estimateUserOperationGas, resolves #752 

## Changes

- Added default function argument to support entryPointWrapper
- Provided entryPointWrapper to UserOpMethodHandler constructor in tests
- Mapped simulateUserOp result to unwrapped type (needs to be verified if the mapping is sufficient for other usages)

## Testing

All tests run through and estimateUserOperationGas now supports the wrapper.

### CODE REVIEW

**Comment by wsdt** [comment]
image
 
 AA integration tests seem to run through as well locally. 
 
 @InoMurko @souradeep-das

**Comment by InoMurko** [comment]
> image
> 
>  
> 
>  AA integration tests seem to run through as well locally. 
> 
>  
> 
>  @InoMurko @souradeep-das 
> 
>  
> 
>  

Yes they do. But they don't use that RPC method.

**Comment by wsdt** [comment]
> > image
> > AA integration tests seem to run through as well locally.
> > @InoMurko @souradeep-das
> 
> Yes they do. But they don't use that RPC method.

it's just an excerpt :-)

**Comment by souradeep-das** [comment]
can we add an integ test for the "estimateUserOperationGas"?

File: packages/boba/bundler/src/UserOpMethodHandler.ts [hunk 1]
```diff
@@ -156,9 +156,15 @@ export class UserOpMethodHandler {
     // todo: checks the existence of parameters, but since we hexlify the inputs, it fails to validate
     await this._validateParameters(deepHexlify(userOp), entryPointInput)
     // todo: validation manager duplicate?
-    const errorResult = await this.entryPoint.callStatic
+    let errorResult = await this.entryPointWrapper.callStatic
```
Comment (InoMurko) [inline packages/boba/bundler/src/UserOpMethodHandler.ts]: there's a comment on line 138: 
```
* TODO: gonna work post bedrock!
```

pls remove it

File: integration-tests/test/eth-l2/boba_aa_bundler.spec.ts [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+      ethers.provider,
+      recipient.signer,
+      config,
+      entryPoint,
+      entryPointWrapper
+    )
+  })
+
+  describe('query rpc calls: eth_estimateUserOperationGas, eth_callUserOperation', async () => {
+    it.only('estimateUserOperationGas should estimate even without eth', async () => {
```
Comment (InoMurko) [inline integration-tests/test/eth-l2/boba_aa_bundler.spec.ts]:
```suggestion
    it('estimateUserOperationGas should estimate even without eth', async () => {
```

File: integration-tests/test/eth-l2/boba_aa_bundler.spec.ts [hunk 3]
```diff
@@ -0,0 +1,207 @@
+import chai from 'chai'
```
Comment (InoMurko) [inline integration-tests/test/eth-l2/boba_aa_bundler.spec.ts:1]: does this test need to be add to bnb and avax?

**Review by InoMurko — APPROVED** [review]

File: ops/docker-compose-side.yml [hunk 4]
```diff
@@ -32,7 +32,6 @@ services:
       dockerfile: ./ops/docker/Dockerfile.packages
       target: gas_oracle
     environment:
-      ADDRESS_MANAGER_ADDRESS: "0x5FbDB2315678afecb367f032d93F642f64180aa3"
```
Comment (wsdt) [inline ops/docker-compose-side.yml:35]: was double defined in develop and broke the compose file

File: ops/docker/Dockerfile.packages [hunk 5]
```diff
@@ -128,6 +128,7 @@ COPY --from=builder /opt/optimism/integration-tests ./integration-tests
 COPY --from=builder /opt/optimism/ops_boba/api ./ops_boba/api
 WORKDIR /opt/optimism/packages/boba
 COPY --from=builder /opt/optimism/packages/boba/account-abstraction ./account-abstraction
+COPY --from=builder /opt/optimism/packages/boba/bundler ./bundler
```
Comment (wsdt) [inline ops/docker/Dockerfile.packages]: needed for integration tests docker

File: packages/boba/bundler/package.json [hunk 6]
```diff
@@ -4,6 +4,13 @@
   "version": "0.2.0",
   "author": "Dror Tirosh",
   "license": "MIT",
+  "main": "dist/src/index.js",
```
Comment (wsdt) [inline packages/boba/bundler/package.json]: needed to be valid package

File: integration-tests/test/alt-l2/boba_aa_bundler.spec.ts [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+    )
+    await accountFactory.deployed()
+    console.log('Account Factory deployed to:', accountFactory.address)
+    await accountFactory.createAccount(env.l2Wallet.address, 0)
+    account = await accountFactory.getAddress(env.l2Wallet.address, 0)
+
+    accountAPI = new SimpleAccountAPI({
+      provider: env.l2Provider,
+      entryPointAddress,
+      entryPointWrapperAddress: entryPointWrapper.address,
```
Comment (souradeep-das) [inline integration-tests/test/alt-l2/boba_aa_bundler.spec.ts]: minor: if we are supplying the accountAddress, we could remove the ePWrapper here and from the test

(applies to the other test too)

File: integration-tests/test/alt-l2/boba_aa_bundler.spec.ts [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+  describe('query rpc calls: eth_estimateUserOperationGas, eth_callUserOperation', async () => {
+    it('estimateUserOperationGas should estimate even without eth', async () => {
+      const op = await accountAPI.createSignedUserOp({
+        target: recipient.address,
+        data: recipient.interface.encodeFunctionData('something', ['hello']),
+      })
+
+      const ret: any = await bundlerProvider.estimateUserOpGas(op)
+
+      // verification gas should be high - it creates this wallet
```
Comment (souradeep-das) [inline integration-tests/test/alt-l2/boba_aa_bundler.spec.ts]: the wallet is created in L95?

File: integration-tests/test/alt-l2/boba_aa_bundler.spec.ts [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+      expect(parseInt(ret.verificationGas, 16)).to.be.closeTo(30000, 100000)
+      // execution should be quite low.
+      // (NOTE: actual execution should revert: it only succeeds because the wallet is NOT deployed yet,
```
Comment (souradeep-das) [inline integration-tests/test/alt-l2/boba_aa_bundler.spec.ts]: is this comment outdated? why couldn't this be executed?

File: integration-tests/package.json [hunk 10]
```diff
@@ -60,13 +60,17 @@
...
     "prom-client": "^14.0.1",
-    "rimraf": "^3.0.2",
     "request-promise-native": "^1.0.5",
-    "node-fetch": "^2.6.1",
-    "ip": "^1.1.5"
+    "rimraf": "^3.0.2"
+  },
+  "dependencies": {
+    "@boba/bundler": "0.2.0",
+    "@boba/bundler_utils": "^0.2.3"
```
Comment (souradeep-das) [inline integration-tests/package.json]: these can be removed now? since we use the bundlerProvider?

File: ops/docker/Dockerfile.packages (hunk 5 above)
Comment (souradeep-das) [reply ops/docker/Dockerfile.packages]: we can remove the bundler package from integ tests now?

File: packages/boba/bundler/src/UserOpMethodHandler.ts [hunk 11]
```diff
@@ -156,9 +155,15 @@ export class UserOpMethodHandler {
...
       .catch((e) => e)
+
+    errorResult = {
+      errorName: errorResult[1][0],
+      errorArgs: errorResult[1],
+    }
+
     if (errorResult.errorName === 'FailedOp') {
       throw new RpcError(
         errorResult.errorArgs.at(-1),
```
Comment (souradeep-das) [inline packages/boba/bundler/src/UserOpMethodHandler.ts]: the parsing of the ePWrapper return needs to change a bit i think according to the contract,
for eg - the FailedOpStatus from return is set to true if 'FailedOp' was returned, and the reason from the same struct can be thrown
(only the L167 and L174 conditions needs to be updated)

**Review by souradeep-das — COMMENTED** [review]
great!

File: packages/boba/bundler/src/UserOpMethodHandler.ts (hunk 11 above)
Comment (wsdt) [reply packages/boba/bundler/src/UserOpMethodHandler.ts]: should be fixed! thanks

**Review by souradeep-das — APPROVED** [review]
Thanks, looks great!

**Pull request merged**